### ISSUE TITLE
[API Proposal]: String return from P/Invoke without free

### ISSUE THREAD

**Issue by dancojocaru2000** [dotnet/runtime#76974]
### Background and motivation

Many libraries automatically free strings returned (for example, on the next call).

Without this feature, an `IntPtr` type must be returned, and then the extern function must be wrapped in a C# function that properly converts the `IntPtr` to a `string` without calling free.

### API Proposal

Add a `bool FreeString` (or similar) property to `DllImportAttribute` and/or `LibraryImportAttribute`, with a default true value to preserve the current behaviour.

### API Usage

```csharp
[DllImport("library", FreeString = false)]
public static extern string ExampleFunction();
```

and/or

```csharp
[LibraryImport("library", FreeString = false)]
public static partial string ExampleFunction();
```

**Comment by teo-tsirpanis** [comment]
You can achieve this with a custom marshaller.

**Comment by dancojocaru2000** [comment]
The documentation for writing custom marshallers is kind of lacking. Even so, there is little benefit to writing a custom marshaller as opposed to just returning `IntPtr` and then writing an extension method like the following:

```csharp
public static class IntPtrToStr {
	public static unsafe string? ToUTF8String(this IntPtr ptr) {
		return Utf8StringMarshaller.ConvertToManaged((byte*)ptr.ToPointer());
	}
}
```

However, I think that having the library free a string instead of passing the responsibility to the caller is common enough that this should be a supported use case by the built in marshaller, as opposed to having someone copy-paste that extension method in all their projects.

**Comment by teo-tsirpanis** [comment]
(small nit on the snippet you sent; you can directly do `(byte*)ptr`)

**Comment by AaronRobinsonMSFT** [comment]
> Add a bool FreeString (or similar) property to DllImportAttribute

The `DllImportAttribute` is a pseudo attribute and is baked into metadata. It isn't possible to add new fields to it without reving the ECMA-33 specification. We would need to add a new attribute that the built-in marshaller respects. However, we aren't interested in adding new features into the built-in marshaller. All new feature work is focused on the `LibraryImport` source generator.

As far as adding this to `LibraryImport`, I am with @teo-tsirpanis and this should be accomplished through a custom marshaller for the string type. See [`CustomMarshallerAttribute`](https://learn.microsoft.com/dotnet/api/system.runtime.interopservices.marshalling.custommarshallerattribute) and the [code sample](https://github.com/dotnet/samples/tree/main/core/interop/source-generation/custom-marshalling).

At present, I don't see a lot of value, but I'm open to seeing if the community thinks otherwise. But do note, this is only an option for `LibraryImport`, adding this feature to the `DllImport` scenario is a non-starter.

**Comment by dancojocaru2000** [comment]
If I'm not mistaken, `LibraryImport` uses source generation. Therefore, would adding this to `LibraryImport` makes it a C# only feature?

**Comment by AaronRobinsonMSFT** [comment]
> If I'm not mistaken, `LibraryImport` uses source generation. Therefore, would adding this to `LibraryImport` make it a C# only feature?

Yes. Are you interested in this feature for a non-C# scenario? If so, what one?

**Comment by lf73** [comment]
I ran into what I believe is caused by this issue related to constant string return when trying to upgrade to LibraryImport.  Currently the return is an IntPtr and then converted to a string via Marshal.PtrToStringAnsi(myIntPtr). 

C Library
`const char *rtlsdr_get_device_name(uint32_t index)`

C# Attempt at using LibraryImport
`        [LibraryImport("librtlsdr",
            EntryPoint = "rtlsdr_get_device_name",
            StringMarshalling = StringMarshalling.Custom,
            StringMarshallingCustomType = typeof(AnsiStringMarshaller))]
        internal static partial string GetDeviceName(uint index);`

When called the process just crashes. No exception. No memory violation error. Having the LibraryImport generated code was very beneficial in tracking down the problem. Stepping into the Finally clause is a call to  Free the memory where it just aborts.

In this situation a LibraryImport optional parameter as suggested would fix the issue. Until then, I will look into making a Custom Marshaller.

**Comment by danmoseley** [comment]
@lf73  I'm not an interop person but just curious why you are using ANSI string marshaling for what is presumably a Linux lib, rather then the usual Utf8.

**Comment by lf73** [comment]
> @lf73 I'm not an interop person but just curious why you are using ANSI string marshaling for what is presumably a Linux lib, rather then the usual Utf8.

@danmoseley   I'm not an expert on C string, so mostly just guessing and trying. The library is using NULL terminated strings. Which I hear is not a great thing security wise, now a days. But this is limited to pulling details about a hardware device, so it's probably okay. Also the hardware device itself may be limited on the string type. If I use UTF-8 then I have to replace a bunch of 0's  ( \0 ) with null.  Using the ANSI string marshalling, it just works perfectly with no additional string manipulation.  This is mostly around where I first have to create a buffer, of say 256 bytes, and let the C library fill it. That's another call than what is above, shared below.  That part has been another really big pain point as really the LibraryImport is not able to provide any coding and I seem to be stuck just using IntPtr's... which under the hood just generates an identical DLLImport statement!

```c#
        [LibraryImport("librtlsdr", EntryPoint = "rtlsdr_get_device_usb_strings")]
        internal static unsafe partial int GetDeviceUsbStrings(uint index,
            nint manufact, nint product, nint serial);

        internal static unsafe int GetTest2(uint index, out string manufact, out string product, out string serial)
        {
            fixed (byte* aptr = new byte[256], bptr = new byte[256], cptr = new byte[256])
            {
                var r = GetDeviceUsbStrings(index, (IntPtr)aptr, (IntPtr)bptr, (IntPtr)cptr);
                manufact = Marshal.PtrToStringAnsi((IntPtr)aptr) ?? "";
                product = Marshal.PtrToStringAnsi((IntPtr)bptr) ?? "";
                serial = Marshal.PtrToStringAnsi((IntPtr)cptr) ?? "";
                return r;
            }
        }
```

**Comment by AaronRobinsonMSFT** [comment]
@lf73 Given the following unmanaged signature:

```c
const char *rtlsdr_get_device_name(uint32_t index)
```

I would define the P/Invoke using `LibraryImport` as follows:
```csharp
[LibraryImport("librtlsdr")]
internal static partial void* rtlsdr_get_device_name(uint index);
```

I would then create a simple wrapper for this scenario.

```csharp
static unsafe string GetDeviceName(uint index)
{
    void* s = rtlsdr_get_device_name(index);
    // Handle null as a special case or just pass it on.
    return Marshal.PtrToStringUTF8((IntPtr)s) ?? "";
}
```

> If I use UTF-8 then I have to replace a bunch of 0's ( \0 ) with null.

On .NET, a null terminator for a UTF-8 or ANSI string are the same value, `\0`. ANSI and UTF-8 are treated the same for .NET on non-Windows platforms.

**Comment by Clockwork-Muse** [comment]
> Which I hear is not a great thing security wise, now a days.

null-terminated strings aren't an issue in and of themselves.
It's the (improper) use of various library functions that do/don't take that into account, or writing your own code that forgets that too.

**Comment by lf73** [comment]
@AaronRobinsonMSFT Thank you very much for the feedback. At first glance I dismissed **void** as either unwanted or nothing expected. After review, I see that **void*** has a special role in returning a pointer that could be of any type. I also appreciate the confirmation that there is no way around having to make a wrapper.

@Clockwork-Muse Thank you for the feedback as well.

**Comment by cyraid** [comment]
Having a 'without free' option would be nice. I'm using a library which returns arrays and other things that you don't have to free, so it'd be nice to have that option.

**Comment by patrickmurray18** [comment]
It's quite shocking that freeing more than potentially necessary was chosen as the default behaviour...

**Comment by tannergooding** [comment]
This is a historical default that can't be changed due to back-compat and largely stems from the expected default behavior of COM and other core scenarios at the time.

The solution is to define a custom marshaller and use the newer LibaryImport source generator feature.
